src/main/java/org/spongepowered/api/text/Text.java

```diff
@@ -0,0 +1,575 @@
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import org.spongepowered.api.text.action.ClickAction;
+import org.spongepowered.api.text.action.HoverAction;
+import org.spongepowered.api.text.action.ShiftClickAction;
+import org.spongepowered.api.text.format.TextColor;
+import org.spongepowered.api.text.format.TextColors;
+import org.spongepowered.api.text.format.TextStyle;
+import org.spongepowered.api.text.format.TextStyles;
+import org.spongepowered.api.text.translation.Translation;
+
+import java.util.Iterator;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents an immutable instance of formatted text that can be displayed on
+ * the client. Each instance consists of content and a list of children texts
+ * appended after the content of this text. The content of the text is available
+ * through one of the subclasses.
+ *
+ * <p>Text is primarily used for sending formatted chat messages to players, but
+ * also in other places like books or signs.</p>
+ *
+ * <p>Text instances can be either directly created through the available
+ * constructor or using the {@link TextBuilder} available through one of the
+ * {@link Texts#builder()} methods (recommended).</p>
+ *
+ * @see Texts#builder()
+ * @see TextBuilder
+ * @see Literal
+ * @see Translatable
+ * @see Selector
+ * @see Score
+ */
+public abstract class Text {
+
+    protected final TextColor color;
+    protected final TextStyle style;
+    protected final ImmutableList<Text> children;
+    protected final Optional<ClickAction<?>> clickAction;
+    protected final Optional<HoverAction<?>> hoverAction;
+    protected final Optional<ShiftClickAction<?>> shiftClickAction;
+
+    protected final Iterable<Text> childrenIterable = new Iterable<Text>() {
+
+        @Override
+        public Iterator<Text> iterator() {
+            return Text.this.children.isEmpty() ? Iterators.singletonIterator(Text.this) : new TextIterator(Text.this);
+        }
+
+    };
+
+    Text() {
+        this(TextColors.NONE, TextStyles.NONE, ImmutableList.<Text>of(), null, null, null);
+    }
+
+    /**
+     * Constructs a new immutable {@link Text} with the specified formatting and
+     * text actions applied.
+     *
+     * @param color The color of the text
+     * @param style The style of the text
+     * @param children The immutable list of children of the text
+     * @param clickAction The click action of the text, or {@code null} for none
+     * @param hoverAction The hover action of the text, or {@code null} for none
+     * @param shiftClickAction The shift click action of the text, or
+     *        {@code null} for none
+     */
+    protected Text(TextColor color, TextStyle style, ImmutableList<Text> children, @Nullable ClickAction<?> clickAction,
```

Comment:
There is the `TextBuilder` for a special reason so I don't see a big reason to provide constructors for convenience in `Text` directly.